compiler/hash-ast-passes/src/visitor.rs

```diff
@@ -600,6 +597,32 @@ impl AstVisitor for SemanticAnalyser<'_> {
     ) -> Result<Self::ParamRet, Self::Error> {
         let _ = walk::walk_param(self, ctx, node);
 
+        if matches!(node.origin, ParamOrigin::Fn) {
+            match self.current_block {
+                // Check that `self` cannot be within a free standing functions
+                BlockOrigin::Root | BlockOrigin::Mod => {
+                    if let Some(name) = node.name.as_ref() && name.is(CORE_IDENTIFIERS.self_i) {
+                        self.append_error(AnalysisErrorKind::SelfInFreeStandingFn, node);
+                    }
+                }
```

Comment:
Mod is valid having self as it is exactly the same as `impl`. Unless we want to change that..

Reply:
What does `self` reference in `mod`?

Reply:
Well, `Foo := struct() ~ mod {..}` is completely the same as `Foo := struct() ~ impl {..}` with the current spec. The only difference is that members of `mod` are private by default, and members of `impl` are public by default. But in both cases writing `self` inside would refer to the struct.

Reply:
From what we have discussed, is this subject to change in the future?

Reply:
No